src/core/providers/ogr/qgsogrprovider.cpp

```diff
@@ -886,16 +886,90 @@ void QgsOgrProvider::loadFields()
   QMutexLocker locker( datasetMutex );
 #endif
 
-  for ( int i = 0; i < fdef.GetFieldCount(); ++i )
+  for ( int fieldIndex = 0; fieldIndex < fdef.GetFieldCount(); ++fieldIndex )
   {
-    OGRFieldDefnH fldDef = fdef.GetFieldDefn( i );
+    OGRFieldDefnH fldDef = fdef.GetFieldDefn( fieldIndex );
     const OGRFieldType ogrType = OGR_Fld_GetType( fldDef );
     const OGRFieldSubType ogrSubType = OGR_Fld_GetSubType( fldDef );
 
     QMetaType::Type varType = QMetaType::Type::UnknownType;
     QMetaType::Type varSubType = QMetaType::Type::UnknownType;
     QgsOgrUtils::ogrFieldTypeToQVariantType( ogrType, ogrSubType, varType, varSubType );
 
+    // Handle special case for OGRFieldType::OFSTJSON which is not necessarily a map.
+    // If subtype is JSON try to load a feature and check if it's
+    // really an object (rather than something else like an array)
+    // fallback to string.
+    if ( ogrType == OFTString && ogrSubType == OFSTJSON )
+    {
+      QRecursiveMutex *layerMutex = nullptr;
+      OGRLayerH ogrLayer = mOgrLayer->getHandleAndMutex( layerMutex );
+      QMutexLocker layerLocker( layerMutex );
+      gdal::ogr_feature_unique_ptr f( OGR_L_GetNextFeature( ogrLayer ) );
+      if ( f )
+      {
+        const char *json = OGR_F_GetFieldAsString( f.get(), fieldIndex );
+        if ( json && json[0] != '\0' )
+        {
+          try
+          {
+            const nlohmann::json json_element = json::parse( json );
+            if ( json_element.is_array() )
+            {
+              bool allStrings = true;
+              bool allInts = true;
+              bool allDoubles = true;
+              for ( const auto &item : json_element )
+              {
+                if ( !item.is_string() )
+                {
+                  allStrings = false;
+                }
+                if ( !item.is_number_integer() )
+                {
+                  allInts = false;
+                }
+                if ( !item.is_number_float() )
```

Comment:
perhaps add a test case for this:
```suggestion
                // We accept integers in mixed list of integer and double, represented as a list of double
                else if ( !item.is_number_float() )
```

Reply:
I have added a test and found out that this code is never reached: homogeneous arrays of numbers and strings are already correctly identified by OGR.

The only cases we need to handle here is when it's an array of mixed types (numbers and strings) or if it contains objects with not  homogeneous values.

I have moved the above logic to the objects identification to make sure we fall back to string when a complex object is found.